plutus-tx/test/laws/Main.hs

```diff
@@ -0,0 +1,533 @@
+      (genN, genD) <- Gen.element [(genIntegerPos, negate <$> genIntegerPos),
+                                   (negate <$> genIntegerPos, genIntegerPos)]
+      Ratio.unsafeRatio <$> genN <*> genD
+
+propAbsRound :: Property
+propAbsRound = property $ do
+  r <- forAllWith ppShow genRational
+  let rounded = Ratio.round r
+  let (n, _) = Ratio.properFraction r
+  assert (Plutus.abs rounded Plutus.>= Plutus.abs n)
+
+propRoundHalf :: Property
+propRoundHalf = property $ do
+  (n, f) <- forAllWith ppShow go
+  let r = Ratio.fromInteger n Plutus.+ f
+  let rounded = Ratio.round r
+  case (signum n, even n) of
+    (-1, False) -> rounded === n Plutus.- Plutus.one
+    (-1, True)  -> rounded === n
+    (0, _)      -> rounded === Plutus.zero
+    (1, False)  -> rounded === n Plutus.+ Plutus.one
+    _           -> rounded === n
+  where
+    go :: Gen (Integer, Plutus.Rational)
+    go = do
+      n <- genInteger
+      f <- case signum n of
+            (-1) -> pure . Ratio.negate $ Ratio.half
+            0    -> Gen.element [Ratio.half, Ratio.negate Ratio.half]
+            _    -> pure Ratio.half
+      pure (n, f)
+
+propRoundLow :: Property
+propRoundLow = property $ do
+  (n, f) <- forAllWith ppShow go
+  let r = Ratio.fromInteger n Plutus.+ f
+  let rounded = Ratio.round r
+  let truncated = Ratio.truncate r
+  rounded === truncated
+  where
+    go :: Gen (Integer, Plutus.Rational)
+    go = do
+      num <- Gen.integral . Range.constant 1 $ 135
+      let f = Ratio.unsafeRatio num 271
+      n <- genInteger
+      case signum n of
+        (-1) -> pure (n, Ratio.negate f)
+        0    -> (n,) <$> Gen.element [f, Ratio.negate f]
+        _    -> pure (n, f)
+
+propRoundHigh :: Property
+propRoundHigh = property $ do
+  (n, f) <- forAllWith ppShow go
+  let r = Ratio.fromInteger n Plutus.+ f
+  let rounded = Ratio.round r
+  let truncated = Ratio.truncate r
+  Plutus.abs rounded === Plutus.abs truncated Plutus.+ Plutus.one
+  where
+    go :: Gen (Integer, Plutus.Rational)
+    go = do
+      num <- Gen.integral . Range.constant 136 $ 270
+      let f = Ratio.unsafeRatio num 271
+      n <- genInteger
+      case signum n of
+        (-1) -> pure (n, Ratio.negate f)
+        0    -> (n,) <$> Gen.element [f, Ratio.negate f]
+        _    -> pure (n, f)
+
+-- Helpers
+
+instance Arg Plutus.Rational where
+  build = via separate (uncurry Ratio.unsafeRatio)
+
+varyRational :: CoGen Plutus.Rational
+varyRational = contramap separate vary
+
+separate :: Plutus.Rational -> (Plutus.Integer, Plutus.Integer)
+separate r = (Ratio.numerator r, Ratio.denominator r)
+
+genRational :: Gen Plutus.Rational
+genRational = do
+  num <- genInteger
+  den <- Gen.integral . Range.linearFrom 100 0 $ 200
+  pure . Ratio.unsafeRatio num $ den
+
+genInteger :: Gen Integer
+genInteger = Gen.integral . Range.linearFrom 0 (-100) $ 100
+
+genIntegerPos :: Gen Integer
+genIntegerPos = Gen.integral . Range.linearFrom 50 1 $ 100
+
+-- This is a hack to ensure coverage.
```

Comment:
Can you explain the hack a bit more for those of us not in the know?

Reply:
This is quite complex, but I'll try and keep it brief. Consider the symmetry property of a binary relation `R`. We have the property that 'if `R(x, y)`, then `R(y,x)`. If some `x` and `y` are not related, we actually don't care what the outcome is.

When testing the symmetry of _equality_, if we generate both `x` and `y` independently, there is a high probability that we will generate non-equal data. Therefore, if we run the test 'blindly', we're not actually testing anything: most of the time, the test will trivially pass because its precondition is not satisfied by the generated data!

`Entangled` (and `Entangled3`) essentially 'hack' the generator to ensure that half the time we generate the same thing twice (or three times), and the other half the time, we generate normally. Combined with liberal use of `cover`, we can ensure that we cover all possibilities. In fact, `cover` gets used in several places, as this ensures that we don't bias our testing to any particular set of preconditions.

Reply:
write it in the code!